compiler/codegen/cg-expr.cpp

```diff
@@ -4706,26 +4706,33 @@ static GenRet codegenGPUKernelLaunch(CallExpr* call, bool is3d) {
   //   - function name
   //   - grid size (1 arg or 3 args)
   //   - block size (1 arg or 3 args)
-  //   - number of kernel parameters (we can probably drop this safely)
   //   - any number of arguments to be passed to the kernel
   //
   // The runtime function needs the kernel parameters to be passed by address.
   // The other arguments to this primitive are passed along directly.
 
   // number of arguments that are not kernel params
-  int nNonKernelParamArgs = is3d ? 8:4;
+  int nNonKernelParamArgs = is3d ? 7:3;
+  int nKernelParamArgs = call->numActuals() - nNonKernelParamArgs;
+
   const char* fn = is3d ? "chpl_gpu_launch_kernel":"chpl_gpu_launch_kernel_flat";
 
   std::vector<GenRet> args;
   int i = 0;
   for_actuals(actual, call) {
-    if (i >= nNonKernelParamArgs) {
+    if (i++ > nNonKernelParamArgs) {
```

Comment:
I'm finding the logic of these conditionals and `i` hard to follow. Can you write the loop so that `i++` is the last statement in the `for_actuals` loop and not in a conditional?